Improved config validation

NoumbissiValere opened the pull request.

NoumbissiValere commented:
ok @nemesisdesign I will squash the commits

NoumbissiValere commented:
@nemesisdesign This PR looks ready. sorry for the delay

atb00ker commented on django_netjsonconfig/tests/test_template.py:
```diff
@@ -263,3 +263,23 @@ def test_duplicate_files_in_template(self):
         else:
             self.fail('ValidationError not raised!')
+
+    def test_variable_substition(self):
```
Not sure, If I understand the purpose of this test.  Is it intended to be this way?

NoumbissiValere replied:
It is there to ensure validation doesn't fail when saving a config with variables. This test ensures that the variables in config are replaced during config validation.  How did you think it was intended to be?

atb00ker replied:
Okay, correct me if I am wrong, the purpose of this is to ensure that the context variable is changed with it's value in the output, right? i.e`obj.config['dns_server']` should have `4.4.4.4` in it? 
Secondly,
```
obj = self.template_model.objects.get(name='test1')
self.assertEqual(obj.name, 'test1')
```
Here, assert Equal can never fail here, either obj.name is `test1` or `self.template_model.objects.get(name='test1')` returns an error and test errors out instead of failing! :smile:

NoumbissiValere replied:
no @atb00ker . the value of `obj.config['dns_server']` is not actually changed. the `default_values` is just use to trick the `schema` to think that the config is correct and to replace the `context` when the `template` is about to be applied to a device. it doesn't replace the `{{dns}}` in the database. `{{dns}}` is still been saved to the database. so in effect we are testing this [line](https://github.com/openwisp/django-netjsonconfig/pull/178/commits/c1554ac4a159dfe99d9b703252bdbb265e465c52#diff-dd64f186cf77b4ab23c7cdcd1cc61cd6R150). Also, if `{{dns}}` is not replaced by `default_values` during validation, the test will fail because the schema rejected the `config` as it was not valid and the object was not created. I don't know if i have explained it well :smile:

atb00ker replied:
Ah, that explains it, I misunderstood the purpose of the test, thanks! :smile:

atb00ker commented on django_netjsonconfig/base/template.py:
```diff
@@ -110,12 +121,13 @@ def clean(self, *args, **kwargs):
-        c.update(super().get_context())
-        return c
+        context.update(super().get_context())
+        context.update(self.default_values)
```
Added the changes in this function, I am fairly certain these changes were lost during rebasing (as it was not done properly done! )

NoumbissiValere replied:
The context was initially added with these defaults value just for consistency as the `id and name` does no have any significant impact. @nemesisdesign has asked it should be removed since context can now be taken from `default_values`

atb00ker replied:
No, I meant mention of `self.default_values` in this function was missing, 
can you please share how did your initial changes look like? :smile:

NoumbissiValere replied:
@atb00ker here is how my initial changes looked like
```
        c = self.default_values or {}
        c.update(super().get_context())
        return c
```

atb00ker replied:
Got it, thanks for the information about the change, i'll revert it! :smile:

NoumbissiValere commented:
Thanks for reviewing. I will look into them

NoumbissiValere commented:
:+1:

NoumbissiValere commented:
> If I add default values to a template and save, this pops-up:
> 
> {
> "vpnserver1": "vpn.testdomain.com"
> }
> 
> Which is the content of NETJSONCONFIG_CONTEXT defined in the settings.py of the test proejct.
> 
> This should not happen, that value should be handled behind the scenes and it should not be saved in every template default_values.

Please i don't understand. where exactly does this pop up? could you explain to me the senario and how you got there?. I just created a template with `default_values` and didn't notice any pop. 
Thanks

atb00ker commented:
@NoumbissiValere in your settings.py, add:
```
NETJSONCONFIG_CONTEXT = {
    'vpnserver1': 'vpn.testdomain.com'
}
```
Then create a template with default values and save it.

Now reopen the default values and you should see:

![image](https://user-images.githubusercontent.com/24985760/83967216-2037fb80-a8dd-11ea-94da-fc79c8f681c1.png)

Notice `'vpnserver1': 'vpn.testdomain.com'` in there!

NoumbissiValere commented:
Ok @atb00ker thank you. Am fixing the PR :smile:

NoumbissiValere commented:
@atb00ker and @nemesisdesign I have addressed every other issue but for the one below:

>**Validation acts weirdly 2**
> Try this buggy template in the advanced mode:
> 
> {
>     "interfaces": [
>         {
>             "type": "bridge",
>             "stp": false,
>             "bridge_members": [
>                 "eth0 eth0.1"
>             ],
>             "name": "lan"
>         }
>     ]
> }
> The JSON validator should tell you that there's an issue, but it doesn't give you any hint.
> 
> The master branch shows the errors correctly.

I have done this on the master branch but still didn't get any report of hints. I also reacted a new local repo with the current master branch but still get the error reporting. tried it equally on `openwisp-controller` master but am not still getting any error hints. Please is there something am lacking? Thanks

NoumbissiValere commented:
> @NoumbissiValere in your settings.py, add:
> 
> ```
> NETJSONCONFIG_CONTEXT = {
>     'vpnserver1': 'vpn.testdomain.com'
> }
> ```
> 
> Then create a template with default values and save it.
> 
> Now reopen the default values and you should see:
> 
> ![image](https://user-images.githubusercontent.com/24985760/83967216-2037fb80-a8dd-11ea-94da-fc79c8f681c1.png)
> 
> Notice `'vpnserver1': 'vpn.testdomain.com'` in there!

This issue was caused by https://github.com/openwisp/django-netjsonconfig/blob/master/django_netjsonconfig/base/template.py#L133
I removed it and replaced it with 
```
def get_context(self):
        return self.default_values or {}
```

NoumbissiValere commented:
Here is the buggy template on `openwisp-controller`
![Uploading Screenshot from 2020-06-08 08-41-26.png…]()

I don't get any error hints
![Screenshot from 2020-06-08 08-41-58](https://user-images.githubusercontent.com/24882630/84004931-3a68ec80-a964-11ea-909f-cf77917fb142.png)

atb00ker commented:
I am looking into it! :smile:

atb00ker commented:
> This issue was caused by https://github.com/openwisp/django-netjsonconfig/blob/master/django_netjsonconfig/base/template.py#L133
I removed it and replaced it with

This doesn't solve the problem though, It just ignores the variable! :-(
I think this needs to be handled at the time of save, I'll look into it as well! :smile:

atb00ker commented:
@NoumbissiValere 
P.S:
```
    def get_context(self):
        context = copy(self.default_values) or {}
        context.update(super().get_context())
        return context
```
solved it.

P.P.S: @NoumbissiValere can you please rebase to `upstream/improved-config-validation` before pushing! :smile:

NoumbissiValere commented:
Sure :smile: I will do further testing to make sure everything is ok with the changes I made and push

atb00ker commented:
@nemesisdesign 
`Validation acts weirdly 1` & `Validation acts weirdly 2` happens because the default values json used was not valid, which brings an issue that the user should be informed when the json is not valid, worked on that! :smile:

NoumbissiValere commented:
> @nemesisdesign
> `Validation acts weirdly 1` & `Validation acts weirdly 2` happens because the default values json used was not valid, which brings an issue that the user should be informed when the json is not valid, worked on that!

@atb00ker the `validation acts weirdly 2` doesn't depend on the `default_values json` it is a functionality which should work even without the `default values`. I have tested thoroughly but could still see the error hints. so we still need to look into that

NoumbissiValere commented:
@atb00ker Please when you have time, could you test that error hints are displayed when you inside a buggy template in advanced mode? I don't see anything but `inspecting` shows they are there. this happens on my browser across other modules which uses `django-netjsonconfig`. below is a screenshot of the output i get from `openwisp-controller` which is thesame with the output i get from this branch on `django-netjsonconfig`. 
![Screenshot from 2020-06-09 07-22-45](https://user-images.githubusercontent.com/24882630/84122066-ca279d00-aa2f-11ea-95af-86dab8b1dce5.png)


I don't know if the error is because of my browser or it is a general problem. could you test it and give me feedback? thanks :smile:

atb00ker commented:
Not sure why it isn't shown on your end, [on my end it's working as expected](https://i.imgur.com/NpXFQJY.png). (Can you try refreshing your browser cache?)

NoumbissiValere commented:
I had not only refresh but also did a hard reload. But since it's working on your end it's ok. I will work on my browser. Thank you for taking our time to test @atb00ker :blush: 
@nemesisdesign could you review it now? All the changes requested have been done.

atb00ker commented on README.rst:
```diff
@@ -50,6 +50,40 @@ Current features
 * **template tags**: tag templates to automate different types of auto-configurations (eg: mesh, WDS, 4G)
 * **simple HTTP resources**: allow devices to automatically download configuration updates
 * **VPN management**: easily create VPN servers and clients
+* **Template variables**: makes it possible to declare context (configuration variables) in template configuration while setting the default values for these declared context in the default values field to bypass validation.
```
@NoumbissiValere 
Should we move the focus of this line from `bypass validation` to `allows setting variable`; i.e from user's perspective.

P.S: I removed some changes that I made wouldn't required now. Please rebase again. :smile:

NoumbissiValere replied:
Ok @atb00ker :smile:
